pkg/tenancy/interface.go

```diff
@@ -0,0 +1,31 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package tenancy
+
+import (
+	"fmt"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/timescale/promscale/pkg/prompb"
+)
+
+const regexOR = "|"
+
+var ErrUnauthorizedTenant = fmt.Errorf("authorized or invalid tenant")
+
+// ReadAuthorizer tells if a read request is allowed to query via Promscale.
+type ReadAuthorizer interface {
+	// ApplyTenantMatcher applies a safety matcher to incoming query matchers. This safety matcher is responsible
+	// from prevent unauthorized query reads from tenants that the incoming query is not supposed to read.
+	ApplyTenantMatcher(ms []*labels.Matcher) []*labels.Matcher
+}
+
+// WriteAuthorizer tells if a write request is authorized to be written.
+type WriteAuthorizer interface {
+	// isAuthorized verifies if the tenant to be inserted is authorized. It returns a ErrUnauthorizedTenant if the tenant is not authorized.
+	isAuthorized(tenantName string) (error, bool)
```

Comment:
Can you clarify the reason about defining a private function in an interface? Doesn't make much sense to me.

Reply:
It's for tests in write_test.go to access the func. The constructors of configs expose in AuthConfig interface and hence is needed. We can get rid of this by exposing with strict types, and using interface only on the receiving var, but I felt using interface at the very constructor level keeps things well abstracted.